### PR TITLE
feat(agw): AGW containerization

### PR DESCRIPTION
**Pull request by rmeleromira**
## Summary

These are the AGW containerization changes.

- [ ] This change is backwards-breaking

### CODE REVIEW

**Comment by themarwhal** [comment]
Is this ready for review or still WIP?

**Review by hcgatewood — COMMENTED** [review]
Please mark as draft while it's still WIP 🙂

**Review by electronjoe — COMMENTED** [review]
Can you review the linting warnings and address where reasonable?

File: lte/gateway/docker/services/build/freediameter.patch [hunk 1]
```diff
@@ -0,0 +1,12 @@
+diff --git a/third_party/build/bin/freediameter_build.sh b/third_party/build/bin/freediameter_build.sh
```
Comment (electronjoe) [inline lte/gateway/docker/services/build/freediameter.patch]: Can you discuss the source of these patches / need in the PR description maybe? Nothing long but so we have context. There are existing patches for some dependencies in Tim's [third-party/build/patches](https://github.com/magma/magma/tree/master/third_party/build/patches) and I would like to know these are not duplicated here.. or if so why they have to be duplicated.

**Review by electronjoe — COMMENTED** [review]
Excited about this work package, Arun!

File: lte/gateway/docker/.env [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2020 The Magma Authors.
```
Comment (arunuke) [inline lte/gateway/docker/.env:1]: With the docker env changes that are pushed in [PR#8771](https://github.com/magma/magma/pull/8771/files), is this file still required?

File: lte/gateway/docker/docker-compose.override.yml [hunk 3]
```diff
@@ -0,0 +1,11 @@
+version: "3.7"
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.override.yml]: lint warnings for the yaml file format

File: lte/gateway/docker/docker-compose.yml [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+
+# Standard logging for each service
+x-logging: &logging_anchor
+  driver: ${LOG_DRIVER}
+
+# Standard volumes mounted
+x-standard-volumes: &volumes_anchor
+  - ${CONFIGS_DEFAULT_VOLUME}:/etc/magma/
+  - ${CONFIGS_TEMPLATES_PATH}:/etc/magma/templates
+  - ${ROOTCA_PATH}:/var/opt/magma/certs/rootCA.pem
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: According to https://magma.github.io/magma/docs/lte/config_agw, the certs are copied to /var/opt/magma/tmp/certs/rootCA.pem. If there are no historical reasons, this should be okay.

File: lte/gateway/docker/docker-compose.yml [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+  - ${CERTS_VOLUME}:/var/opt/magma/certs
+  - ${CONFIGS_OVERRIDE_VOLUME}:/var/opt/magma/configs
+  - ${SNOWFLAKE_PATH}:/etc/snowflake
+
+x-generic-service: &service
+  volumes: *volumes_anchor
+  logging: *logging_anchor
+  restart: always
+  network_mode: host
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: This should be okay for the current iteration since it attempts to bring up containerized AGW in a 1:1 fashion similar to package based AGW. But depending on the changes happening in pipelined and sessiond that will allow multiple AGWs on the same host, this would have to change.

File: lte/gateway/docker/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    <<: *ltecservice
+    container_name: oai_mme
+    ulimits:
+      core: -1
+    security_opt:
+      - seccomp:unconfined
+    privileged: true
+    command: /usr/local/bin/oai_mme -c /var/opt/magma/tmp/mme.conf -s /var/opt/magma/tmp/spgw.conf
+  #envoy_controller:
+  #  <<: *cservice
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: Removing old code for readability

File: lte/gateway/docker/docker-compose.yml [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+  - ${SNOWFLAKE_PATH}:/etc/snowflake
+
+x-generic-service: &service
+  volumes: *volumes_anchor
+  logging: *logging_anchor
+  restart: always
+  network_mode: host
+
+# Generic python anchor to avoid repetition for agw python services
+x-agw-python-service: &pyservice
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: General feedback on all services - a label needs to be included for all version information or a UUID (indicating which cluster/service this container instance belongs to), represented by a variable that can be switched based on the version that is being built in the dockerfile.

File: lte/gateway/docker/docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,232 @@
+---
+version: "3.7"
+
+# Standard logging for each service
+x-logging: &logging_anchor
+  driver: ${LOG_DRIVER}
+
+# Standard volumes mounted
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: Since these volumes only present values to the container, should these be read-only?

File: lte/gateway/docker/docker-compose.yml [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+x-standard-volumes: &volumes_anchor
+  - ${CONFIGS_DEFAULT_VOLUME}:/etc/magma/
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: Should these configuration files be in the "configs" sub-section instead of volumes?

File: lte/gateway/docker/docker-compose.yml [hunk 10]
```diff
@@ -0,0 +1,232 @@
+---
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: compose.yaml over docker-compose.yml perhaps?

File: lte/gateway/docker/docker-compose.yml [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    <<: *pyservice
+    container_name: subscriberdb
+    command: /usr/bin/env python3 -m magma.subscriberdb.main
+  policydb:
+    <<: *pyservice
+    container_name: policydb
+    depends_on:
+      - redis
+    command: /usr/bin/env python3 -m magma.policydb.main
+  pipelined:
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: Should pipelined be allocated a number of base CPUs to be available for the process?

File: lte/gateway/docker/docker-compose.yml [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+  #   <<: *pyservice
+  #   container_name: mobilityd
+
+  # Prototype ??
+  # ovs:
+  #   image: openvswitch
+  #   container_name: ovs
+  redis:
+    <<: *pyservice
+    container_name: redis
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: Is redis not expected to be persistent (thereby needing a volume)? PS : Updating this comment - Although Redis saves data by snapshotting it locally, policydb continues to stream data from cloud on restarts. We should be okay with this approach for now.

File: lte/gateway/docker/docker-compose.yml [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+  magmad:
+    <<: *pyservice
+    container_name: magmad
+    environment:
+      DOCKER_REGISTRY: ${DOCKER_REGISTRY}
+      DOCKER_USERNAME: ${DOCKER_USERNAME}
+      DOCKER_PASSWORD: ${DOCKER_PASSWORD}
+    command: /usr/bin/env python3 -m magma.magmad.main
+  mobilityd:
+    <<: *pyservice
```
Comment (arunuke) [inline lte/gateway/docker/docker-compose.yml]: Some services have a healthcheck, while others seemingly don't. How can we rationalize this?

**Review by arunuke — CHANGES_REQUESTED** [review]
For review comments that are new feature requests, please file a new issue that will add the functionality. This would ensure that this revision is unblocked while newer features are added in the next sprint.

File: lte/gateway/docker/example/subscriberdb.yml [hunk 14]
```diff
@@ -0,0 +1,58 @@
+---
+#
```
Comment (arunuke) [inline lte/gateway/docker/example/subscriberdb.yml]: Are all example files needed?

File: lte/gateway/docker/services/build/freediameter.patch (hunk 1 above)
Comment (arunuke) [reply lte/gateway/docker/services/build/freediameter.patch]: Following up on this comment - redundant files will be dropped from the PR after discussion with @rmeleromira

File: third_party/build/bin/freediameter_build.sh [hunk 15]
```diff
@@ -61,7 +61,7 @@ git checkout "${GIT_COMMIT}"
```
Comment (arunuke) [inline third_party/build/bin/freediameter_build.sh]: Will this file/changes be skipped?

File: third_party/build/bin/gnutls_build.sh [hunk 16]
```diff
@@ -61,7 +61,7 @@ cd ${WORK_DIR}
 wget http://mirrors.dotsrc.org/gcrypt/gnutls/v3.1/gnutls-$PKGVERSION.tar.xz
```
Comment (arunuke) [inline third_party/build/bin/gnutls_build.sh]: Will these changes be skipped?

File: third_party/build/bin/grpc_build.sh [hunk 17]
```diff
@@ -22,7 +22,7 @@ set -e
 SCRIPT_DIR="$(dirname "$(realpath "$0")")"
```
Comment (arunuke) [inline third_party/build/bin/grpc_build.sh]: Will these changes be skipped?

File: third_party/build/bin/nettle_build.sh [hunk 18]
```diff
@@ -19,7 +19,7 @@ set -e
 SCRIPT_DIR="$(dirname "$(realpath "$0")")"
```
Comment (arunuke) [inline third_party/build/bin/nettle_build.sh]: Will these changes be skipped?

**Review by arunuke — CHANGES_REQUESTED** [review]

File: lte/gateway/docker/.env (hunk 2 above)
Comment (ymasmoudi) [reply lte/gateway/docker/.env:1]: no, we should delete this one

File: lte/gateway/docker/docker-compose.yml (hunk 4 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yml]: Actually, feg and cwag, xwfm expect the rootca here /var/opt/magma/certs/rootCA.pem
But yes, it does not matter as long as the control proxy config matches.

File: lte/gateway/docker/example/subscriberdb.yml (hunk 14 above)
Comment (ymasmoudi) [reply lte/gateway/docker/example/subscriberdb.yml]: no, I removed them in a previous commit.
we should rely on default config for now

File: lte/gateway/docker/docker-compose.yaml [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+    privileged: true
+    command: >
+      bash -c "/usr/bin/ovs-vsctl --all destroy Flow_Sample_Collector_Set &&
+        /usr/bin/ovs-vsctl set bridge gtp_br0 protocols=OpenFlow10,OpenFlow13,OpenFlow14 other-config:disable-in-band=true &&
+        /usr/bin/ovs-vsctl set-controller gtp_br0 tcp:127.0.0.1:6633 tcp:127.0.0.1:6654 &&
+        /usr/bin/ovsdb-client dump Controller _uuid|tail -n +4 | while read id ; do \
+          /usr/bin/ovs-vsctl set Controller $${id} inactivity_probe=300 ; done &&
+        /usr/bin/ovs-vsctl set-fail-mode gtp_br0 secure &&
+        /usr/bin/ovs-vsctl set-manager ptcp:6640 &&
+        /usr/bin/env python3 -m magma.pipelined.main "
```
Comment (pshelar) [inline lte/gateway/docker/docker-compose.yaml]: we need to refactor this in a script and call it from here, otherwise it will drift away from package based deployment.

File: lte/gateway/docker/mme/configs/spgw.conf [hunk 20]
```diff
@@ -0,0 +1,92 @@
+################################################################################
```
Comment (pshelar) [inline lte/gateway/docker/mme/configs/spgw.conf:1]: why this find is added to repo? this is auto generated file.

File: third_party/build/bin/freediameter_build.sh [hunk 21]
```diff
@@ -61,7 +61,7 @@ git checkout "${GIT_COMMIT}"
 
 patch -p1 < $MAGMA_ROOT/lte/gateway/c/oai/patches/0001-opencoord.org.freeDiameter.patch
 
-awk '{if (/^DISABLE_SCTP/) gsub(/OFF/, "ON"); print}' CMakeCache.txt > tmp && mv tmp CMakeCache.txt
+awk '{if (/^DISABLE_SCTP/) gsub(/OFF/, "ON"); print}' build/CMakeCache.txt > tmp && mv tmp build/CMakeCache.txt
```
Comment (pshelar) [inline third_party/build/bin/freediameter_build.sh]: why are these build script changed? containerized AGW can use already built packages for these dependency

File: third_party/build/bin/grpc_build.sh (hunk 17 above)
Comment (pshelar) [inline third_party/build/bin/grpc_build.sh]: same here.

File: third_party/build/bin/nettle_build.sh (hunk 18 above)
Comment (pshelar) [inline third_party/build/bin/nettle_build.sh]: please add comment for every change you are making outside of docker build script.

**Review by pshelar — CHANGES_REQUESTED** [review]
can you merge all patches and rebase it on latest master.

File: third_party/build/bin/nettle_build.sh (hunk 18 above)
Comment (ymasmoudi) [reply third_party/build/bin/nettle_build.sh]: removed

File: third_party/build/bin/grpc_build.sh (hunk 17 above)
Comment (ymasmoudi) [reply third_party/build/bin/grpc_build.sh]: removed

File: third_party/build/bin/freediameter_build.sh (hunk 21 above)
Comment (ymasmoudi) [reply third_party/build/bin/freediameter_build.sh]: I think it was just a nit by ramon. removed.

File: third_party/build/bin/grpc_build.sh (hunk 17 above)
Comment (ymasmoudi) [reply third_party/build/bin/grpc_build.sh]: removed

File: third_party/build/bin/gnutls_build.sh (hunk 16 above)
Comment (ymasmoudi) [reply third_party/build/bin/gnutls_build.sh]: removed

File: third_party/build/bin/freediameter_build.sh (hunk 15 above)
Comment (ymasmoudi) [reply third_party/build/bin/freediameter_build.sh]: removed

File: lte/gateway/docker/docker-compose.yml (hunk 10 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yml]: fixed

File: lte/gateway/docker/docker-compose.yml (hunk 9 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yml]: these are default config copied from repo.

File: lte/gateway/docker/docker-compose.override.yml (hunk 3 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.override.yml]: fixed

File: lte/gateway/docker/mme/configs/spgw.conf (hunk 20 above)
Comment (ymasmoudi) [reply lte/gateway/docker/mme/configs/spgw.conf:1]: tracked in issue #9350

File: lte/gateway/docker/docker-compose.yaml (hunk 19 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yaml]: tracked here in issue #9351

File: lte/gateway/docker/docker-compose.yml (hunk 8 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yml]: some them does not actually, /var/run and /tmp, /var/opt/magma/{configs, certs} are all updated by containers

File: lte/gateway/docker/docker-compose.yml (hunk 11 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yml]: we should let that as a customisation for each agw. today we already allocate cpus for ovs, and we don't have visibility on how many left we have.

File: lte/gateway/docker/docker-compose.yml (hunk 13 above)
Comment (ymasmoudi) [reply lte/gateway/docker/docker-compose.yml]: I added more health check for directoryd, policydb and pipelined. 
we will need to update the others as well.
this is tracked here in issue #9352

File: third_party/build/bin/nettle_build.sh (hunk 18 above)
Comment (ymasmoudi) [reply third_party/build/bin/nettle_build.sh]: removed

File: lte/gateway/docker/services/build/freediameter.patch (hunk 1 above)
Comment (ymasmoudi) [reply lte/gateway/docker/services/build/freediameter.patch]: removed

File: lte/gateway/docker/docker-compose.yaml [hunk 22]
```diff
@@ -0,0 +1,235 @@
...
+             /usr/bin/redis-server /var/opt/magma/tmp/redis.conf --daemonize no &&
+             /usr/bin/redis-cli shutdown"
+
+  directoryd:
+    <<: *pyservice
+    container_name: directoryd
+    depends_on:
+      - redis
+    healthcheck:
+      test: ["CMD", "nc", "-zv", "localhost","50067"]
```
Comment (electronjoe) [inline lte/gateway/docker/docker-compose.yaml]: nit: missing a space

File: lte/gateway/docker/docker-compose.yaml [hunk 23]
```diff
@@ -0,0 +1,235 @@
...
+      - redis
+    command: /usr/bin/env python3 -m magma.state.main
+
+  policydb:
+    <<: *pyservice
+    container_name: policydb
+    depends_on:
+      - redis
+    healthcheck:
+      test: ["CMD", "nc", "-zv", "localhost","50068"]
```
Comment (electronjoe) [inline lte/gateway/docker/docker-compose.yaml]: nit: missing a space between comma separated items

File: lte/gateway/docker/docker-compose.yaml [hunk 24]
```diff
@@ -0,0 +1,235 @@
...
+    security_opt:
+      - seccomp:unconfined
+    privileged: true
+    environment:
+      MAGMA_PRINT_GRPC_PAYLOAD: 1
+    cap_add:
+      - NET_ADMIN
+      - NET_RAW
+    healthcheck:
+      test: ["CMD", "nc", "-zv", "localhost","50063"]
```
Comment (electronjoe) [inline lte/gateway/docker/docker-compose.yaml]: nit: missing a space between comma separated elements

File: lte/gateway/docker/docker-compose.yaml [hunk 25]
```diff
@@ -0,0 +1,235 @@
...
+      core: -1
+    security_opt:
+      - seccomp:unconfined
+    privileged: true
+    environment:
+      MAGMA_PRINT_GRPC_PAYLOAD: 1
+    depends_on:
+      - directoryd
+    healthcheck:
+      test: ["CMD", "nc", "-zv", "localhost","50065"]
```
Comment (electronjoe) [inline lte/gateway/docker/docker-compose.yaml]: nit: missing a space between comma separated elements

**Review by electronjoe — APPROVED** [review]

**Comment by ymasmoudi** [comment]
@arun-magma  could you approve as well since you requested changes ?

File: lte/gateway/docker/mme/configs/spgw.conf [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+    };
+
+    # DNS address communicated to UEs
+    DEFAULT_DNS_IPV4_ADDRESS     = "8.8.8.8";
+    DEFAULT_DNS_SEC_IPV4_ADDRESS = "8.8.4.4";
+    DNS_SERVER_IPV6_ADDRESS      = "2001:4860:4860:0:0:0:0:8888";
+    P_CSCF_IPV4_ADDRESS          = "172.27.23.150";
+    P_CSCF_IPV6_ADDRESS          = "2a12:577:9941:f99c:0002:0001:c731:f114";
+
+    # Non standard feature, normally should be set to "no", but you may need to set to yes for UE that do not explicitly request a PDN address through NAS signalling
```
Comment (arunuke) [inline lte/gateway/docker/mme/configs/spgw.conf]: minor typo

File: lte/gateway/docker/mme/configs/spgw.conf [hunk 27]
```diff
@@ -0,0 +1,92 @@
...
+    FORCE_PUSH_PROTOCOL_CONFIGURATION_OPTIONS = "no";                           # STRING, {"yes", "no"}.
+    UE_MTU                                    = 1400         # MTU - (extended GTPv1 hdr(16 Bytes) + UDP hdr(8) -IPv4(20) hdr + additonal bytes(56)) INTEGER
```
Comment (arunuke) [inline lte/gateway/docker/mme/configs/spgw.conf]: minor typo - additional

File: lte/gateway/docker/publish.sh [hunk 28]
```diff
@@ -0,0 +1,9 @@
+REGISTRY=$1
```
Comment (arunuke) [inline lte/gateway/docker/publish.sh:13]: using ! to specify shell

File: lte/gateway/docker/publish.sh [hunk 29]
```diff
@@ -0,0 +1,9 @@
...
+VERSION=latest
+publish (){
+  docker tag agw_gateway_$1:latest ${REGISTRY}gateway_$1:${VERSION}
```
Comment (arunuke) [inline lte/gateway/docker/publish.sh]: usage of double quotes

**Review by arunuke — APPROVED** [review]
Added a few minor comments based on output from automated tools. To be targeted in the refactor.

**Comment by arunuke** [comment]
> @arun-magma could you approve as well since you requested changes ?

If you were looking for me (@arunuke :)), I added my comments and approved. Might need an issue for some of the refactors as suggested by the automated tools.

File: lte/gateway/docker/mme/configs/spgw.conf (hunk 27 above)
Comment (ymasmoudi) [reply lte/gateway/docker/mme/configs/spgw.conf]: fixed

File: lte/gateway/docker/mme/configs/spgw.conf (hunk 26 above)
Comment (ymasmoudi) [reply lte/gateway/docker/mme/configs/spgw.conf]: fixed

File: lte/gateway/docker/publish.sh (hunk 28 above)
Comment (ymasmoudi) [reply lte/gateway/docker/publish.sh:13]: fixed

File: lte/gateway/docker/publish.sh (hunk 29 above)
Comment (ymasmoudi) [reply lte/gateway/docker/publish.sh]: fixed

File: lte/gateway/docker/mme/configs/mme.conf [hunk 30]
```diff
@@ -0,0 +1,296 @@
+################################################################################
```
Comment (pshelar) [inline lte/gateway/docker/mme/configs/mme.conf:1]: can you create sym-link here.

File: lte/gateway/docker/mme/configs/spgw.conf [hunk 31]
```diff
@@ -0,0 +1,92 @@
+################################################################################
+# Copyright 2020 The Magma Authors.
+
```
Comment (pshelar) [inline lte/gateway/docker/mme/configs/spgw.conf:3]: same as above.

**Review by pshelar — COMMENTED** [review]
This looks pretty close.

File: lte/gateway/docker/mme/configs/mme.conf (hunk 30 above)
Comment (ymasmoudi) [reply lte/gateway/docker/mme/configs/mme.conf:1]: @pshelar sym-link  ? this file is not duplicated. 
the repo has only a template file which we don't support yet. auto config generation for mme will follow up in another PR. 
issue #9350

File: lte/gateway/docker/mme/configs/spgw.conf (hunk 31 above)
Comment (ymasmoudi) [reply lte/gateway/docker/mme/configs/spgw.conf:3]: same as above :)

File: lte/gateway/docker/mme/configs/mme.conf (hunk 30 above)
Comment (pshelar) [reply lte/gateway/docker/mme/configs/mme.conf:1]: ok

**Review by pshelar — APPROVED** [review]
LGTM
